# Import of libraries


In [118]:
#libraries
import pandas as pd
import seaborn as sns
import matplotlib as plt
import numpy as np
from sklearn.preprocessing import LabelEncoder

# Dataset "transactions_train.csv"


I decided to import this dataset only for the analysis of a value in another dataset and not for working on missing data since there are not

## Import of dataset

In [119]:
#import of the csv file
transactions = pd.read_csv('Data/transactions_train.csv')
transactions = pd.DataFrame(transactions)

In [120]:
transactions

t_dat                                        customer_id  \
0         2018-09-20  000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...   
1         2018-09-20  000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...   
2         2018-09-20  00007d2de826758b65a93dd24ce629ed66842531df6699...   
3         2018-09-20  00007d2de826758b65a93dd24ce629ed66842531df6699...   
4         2018-09-20  00007d2de826758b65a93dd24ce629ed66842531df6699...   
...              ...                                                ...   
31788319  2020-09-22  fff2282977442e327b45d8c89afde25617d00124d0f999...   
31788320  2020-09-22  fff2282977442e327b45d8c89afde25617d00124d0f999...   
31788321  2020-09-22  fff380805474b287b05cb2a7507b9a013482f7dd0bce0e...   
31788322  2020-09-22  fff4d3a8b1f3b60af93e78c30a7cb4cf75edaf2590d3e5...   
31788323  2020-09-22  fffef3b6b73545df065b521e19f64bf6fe93bfd450ab20...   

          article_id     price  sales_channel_id  
0          663713001  0.050831                 2  
1          541518023  0.030492                 2  
2          505221004  0.015237                 2  
3          685687003  0.016932                 2  
4          685687004  0.016932                 2  
...              ...       ...               ...  
31788319   929511001  0.059305                 2  
31788320   891322004  0.042356                 2  
31788321   918325001  0.043203                 1  
31788322   833459002  0.006763                 1  
31788323   898573003  0.033881                 2  

[31788324 rows x 5 columns]

# Dataset "articles.csv"

## Import of the dataset

In [121]:
#import of the csv file
articles = pd.read_csv('Data/articles.csv')
articles = pd.DataFrame(articles)

In [122]:
articles

article_id  product_code               prod_name  product_type_no  \
0        108775015        108775               Strap top              253   
1        108775044        108775               Strap top              253   
2        108775051        108775           Strap top (1)              253   
3        110065001        110065       OP T-shirt (Idro)              306   
4        110065002        110065       OP T-shirt (Idro)              306   
...            ...           ...                     ...              ...   
105537   953450001        953450  5pk regular Placement1              302   
105538   953763001        953763       SPORT Malaga tank              253   
105539   956217002        956217         Cartwheel dress              265   
105540   957375001        957375        CLAIRE HAIR CLAW               72   
105541   959461001        959461            Lounge dress              265   

       product_type_name  product_group_name  graphical_appearance_no  \
0               Vest top  Garment Upper body                  1010016   
1               Vest top  Garment Upper body                  1010016   
2               Vest top  Garment Upper body                  1010017   
3                    Bra           Underwear                  1010016   
4                    Bra           Underwear                  1010016   
...                  ...                 ...                      ...   
105537             Socks      Socks & Tights                  1010014   
105538          Vest top  Garment Upper body                  1010016   
105539             Dress   Garment Full body                  1010016   
105540         Hair clip         Accessories                  1010016   
105541             Dress   Garment Full body                  1010016   

       graphical_appearance_name  colour_group_code colour_group_name  ...  \
0                          Solid                  9             Black  ...   
1                          Solid                 10             White  ...   
2                         Stripe                 11         Off White  ...   
3                          Solid                  9             Black  ...   
4                          Solid                 10             White  ...   
...                          ...                ...               ...  ...   
105537           Placement print                  9             Black  ...   
105538                     Solid                  9             Black  ...   
105539                     Solid                  9             Black  ...   
105540                     Solid                  9             Black  ...   
105541                     Solid                 11         Off White  ...   

          department_name index_code        index_name index_group_no  \
0            Jersey Basic          A        Ladieswear              1   
1            Jersey Basic          A        Ladieswear              1   
2            Jersey Basic          A        Ladieswear              1   
3          Clean Lingerie          B  Lingeries/Tights              1   
4          Clean Lingerie          B  Lingeries/Tights              1   
...                   ...        ...               ...            ...   
105537          Socks Bin          F          Menswear              3   
105538             Jersey          A        Ladieswear              1   
105539             Jersey          A        Ladieswear              1   
105540  Small Accessories          D           Divided              2   
105541             Jersey          A        Ladieswear              1   

        index_group_name section_no            section_name garment_group_no  \
0             Ladieswear         16  Womens Everyday Basics             1002   
1             Ladieswear         16  Womens Everyday Basics             1002   
2             Ladieswear         16  Womens Everyday Basics             1002   
3             Ladieswear         61         Womens Lingerie             

## Feature engeneering: preliminary data analysis


In [123]:
#check about the correctness of the information found in the past analysis (column and numbers of missing values)
articles.isna().sum()

article_id                        0
product_code                      0
prod_name                         0
product_type_no                   0
product_type_name                 0
product_group_name                0
graphical_appearance_no           0
graphical_appearance_name         0
colour_group_code                 0
colour_group_name                 0
perceived_colour_value_id         0
perceived_colour_value_name       0
perceived_colour_master_id        0
perceived_colour_master_name      0
department_no                     0
department_name                   0
index_code                        0
index_name                        0
index_group_no                    0
index_group_name                  0
section_no                        0
section_name                      0
garment_group_no                  0
garment_group_name                0
detail_desc                     416
dtype: int64

In [124]:
#type of the column with missing values
articles_data_types = articles.dtypes
data_type_detail_desc = articles_data_types['detail_desc']
data_type_detail_desc

dtype('O')

From this analysis we can undestand that the missing values in the column "detail_desc" have the type that is an object and we have 416 missing values.
Since is not a numerical feature but a categorical one (is an object that contains strings), it is not possible to substitute the missing values with the replacement based on the other values but is necessary to use another technique.

## Analysis on the missing values

In [125]:
from collections import Counter
#ids of the products with missing values
null_description_id = articles[articles['detail_desc'].isnull()]['article_id']

#type of the products with missing values
null_type_product = articles[articles['detail_desc'].isnull()]['product_type_name']

#count of products by name
element_count_name_product = Counter(null_type_product)

#count of product with id
element_count_id = Counter(null_description_id)

In [126]:
element_count_name_product

Counter({'Sweater': 31,
         'Boots': 3,
         'Underwear bottom': 29,
         'Ballerinas': 31,
         'T-shirt': 55,
         'Sandals': 12,
         'Dress': 44,
         'Top': 11,
         'Hoodie': 7,
         'Trousers': 29,
         'Sneakers': 31,
         'Bra': 11,
         'Vest top': 4,
         'Jacket': 6,
         'Swimsuit': 11,
         'Sunglasses': 3,
         'Hat/brim': 1,
         'Shorts': 10,
         'Swimwear bottom': 11,
         'Skirt': 9,
         'Shirt': 8,
         'Blouse': 15,
         'Socks': 3,
         'Coat': 5,
         'Slippers': 1,
         'Hair string': 1,
         'Underwear body': 1,
         'Bodysuit': 1,
         'Earring': 4,
         'Sarong': 1,
         'Other shoe': 5,
         'Dungarees': 1,
         'Leggings/Tights': 1,
         'Flip flop': 1,
         'Bracelet': 2,
         'Other accessories': 1,
         'Cardigan': 5,
         'Bag': 1,
         'Scarf': 1,
         'Hair/alice band': 1,
         'Flat shoe': 

In [127]:
element_count_id

Counter({351332007: 1,
         420049002: 1,
         420049003: 1,
         426199002: 1,
         426199010: 1,
         426199018: 1,
         426199019: 1,
         426199021: 1,
         426199027: 1,
         426199028: 1,
         426199029: 1,
         469039001: 1,
         469039019: 1,
         469039021: 1,
         469039024: 1,
         469039028: 1,
         469039032: 1,
         469039033: 1,
         469039034: 1,
         469039035: 1,
         469039040: 1,
         469039047: 1,
         493438001: 1,
         493438002: 1,
         493438005: 1,
         493438012: 1,
         493438013: 1,
         493438014: 1,
         493438015: 1,
         493438018: 1,
         493438019: 1,
         493438020: 1,
         493438021: 1,
         493438025: 1,
         493438027: 1,
         507883002: 1,
         507883007: 1,
         507883008: 1,
         507883009: 1,
         507883011: 1,
         507883014: 1,
         507883015: 1,
         507883018: 1,
         51

In [128]:
#find of number of transactions with articles without description
k = transactions['article_id'].isin(null_description_id)
number_transactions_per_article = dict(transactions['article_id'][k].value_counts())
number_transactions_per_article

{507883009: 8318,
 742916002: 4524,
 507883014: 4516,
 620081001: 3915,
 469039019: 3875,
 742916003: 3378,
 426199002: 3359,
 684824003: 3175,
 684824002: 2458,
 663518004: 2413,
 663518002: 2407,
 742916001: 2281,
 719463001: 1783,
 661147004: 1764,
 567993016: 1724,
 686631001: 1613,
 684824006: 1482,
 684824001: 1402,
 681156002: 1360,
 567993015: 1319,
 493438027: 1271,
 687722001: 1186,
 426199021: 1155,
 685150002: 1098,
 620081010: 1045,
 586282001: 1039,
 686631002: 1031,
 719463002: 1013,
 620188001: 925,
 686631003: 923,
 493438025: 912,
 661147008: 891,
 719463003: 879,
 567993027: 853,
 620081008: 751,
 567993017: 747,
 620081006: 741,
 610543011: 690,
 567993042: 659,
 507883018: 643,
 599716001: 632,
 469039034: 630,
 688037004: 630,
 620081007: 592,
 637268009: 591,
 637268006: 575,
 598795017: 574,
 733668001: 573,
 693500002: 558,
 580331004: 543,
 567993043: 524,
 743641001: 521,
 578311010: 491,
 567993040: 485,
 507883015: 456,
 562480004: 452,
 586265002: 437,
 75

In [129]:
#sum of all the articles bought that don't have a description
things_buy = number_transactions_per_article.values()

sum_articles = sum(things_buy)
print(sum_articles)

115027


In [130]:
#percentage of transactions of an article without the description
print((sum_articles/len(transactions))*100)

0.3618529872792287


# Dataset "customers.csv"

## Import of data

In [131]:
#import of the csv file
customers = pd.read_csv('Data/customers.csv')
customers = pd.DataFrame(customers)

In [132]:
customers

customer_id   FN  Active  \
0        00000dbacae5abe5e23885899a1fa44253a17956c6d1c3...  NaN     NaN   
1        0000423b00ade91418cceaf3b26c6af3dd342b51fd051e...  NaN     NaN   
2        000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...  NaN     NaN   
3        00005ca1c9ed5f5146b52ac8639a40ca9d57aeff4d1bd2...  NaN     NaN   
4        00006413d8573cd20ed7128e53b7b13819fe5cfc2d801f...  1.0     1.0   
...                                                    ...  ...     ...   
1371975  ffffbbf78b6eaac697a8a5dfbfd2bfa8113ee5b403e474...  NaN     NaN   
1371976  ffffcd5046a6143d29a04fb8c424ce494a76e5cdf4fab5...  NaN     NaN   
1371977  ffffcf35913a0bee60e8741cb2b4e78b8a98ee5ff2e6a1...  1.0     1.0   
1371978  ffffd7744cebcf3aca44ae7049d2a94b87074c3d4ffe38...  1.0     1.0   
1371979  ffffd9ac14e89946416d80e791d064701994755c3ab686...  NaN     NaN   

        club_member_status fashion_news_frequency   age  \
0                   ACTIVE                   NONE  49.0   
1                   ACTIVE                   NONE  25.0   
2                   ACTIVE                   NONE  24.0   
3                   ACTIVE                   NONE  54.0   
4                   ACTIVE              Regularly  52.0   
...                    ...                    ...   ...   
1371975             ACTIVE                   NONE  24.0   
1371976             ACTIVE                   NONE  21.0   
1371977             ACTIVE              Regularly  21.0   
1371978             ACTIVE              Regularly  18.0   
1371979         PRE-CREATE                   NONE  65.0   

                                               postal_code  
0        52043ee2162cf5aa7ee79974281641c6f11a68d276429a...  
1        2973abc54daa8a5f8ccfe9362140c63247c5eee03f1d93...  
2        64f17e6a330a85798e4998f62d0930d14db8db1c054af6...  
3        5d36574f52495e81f019b680c843c443bd343d5ca5b1c2...  
4        25fa5ddee9aac01b35208d01736e57942317d756b32ddd...  
...                                                    ...  
1371975  7aa399f7e669990daba2d92c577b52237380662f36480b...  
1371976  3f47f1279beb72215f4de557d950e0bfa73789d24acb5e...  
1371977  4563fc79215672cd6a863f2b4bf56b8f898f2d96ed590e...  
1371978  8892c18e9bc3dca6aa4000cb8094fc4b51ee8db2ed14d7...  
1371979  0a1a03306fb2f62164c2a439b38c0caa64b40deaae8687...  

[1371980 rows x 7 columns]

## Feature engeneering: preliminary analysis on data



In [133]:
#check about the correctness of the information found in the past analysis (column and numbers of missing values)
customers.isna().sum()

customer_id                    0
FN                        895050
Active                    907576
club_member_status          6062
fashion_news_frequency     16009
age                        15861
postal_code                    0
dtype: int64

In [134]:
#type of the column with missing values
customers_data_types = customers.dtypes
FN_type = customers_data_types['FN']
active_type = customers_data_types['Active']
club_member_status_type = customers_data_types['club_member_status']
fashion_news_frequency_type = customers_data_types['fashion_news_frequency']
age_type = customers_data_types['age']

In [135]:
FN_type

dtype('float64')

In [136]:
active_type

dtype('float64')

In [137]:
age_type

dtype('float64')

Since these values are a float64, so are a numerical value, is possible to modify the missing data based on statistical parameters of the values that are not missing or using other techniques

In [138]:
club_member_status_type

dtype('O')

In [139]:
fashion_news_frequency_type

dtype('O')

Since these values are all objects is necessary to use different techniquest that are not the ones used for the substitution of missing values with the numerical ones

## Feature enegeneering on a numerical variable: age, FN and Active

### age

In [140]:
#first I want to discover which are the possible values of the variable age
age_values = customers['age'].unique()
age_values.sort()
age_values

array([16., 17., 18., 19., 20., 21., 22., 23., 24., 25., 26., 27., 28.,
       29., 30., 31., 32., 33., 34., 35., 36., 37., 38., 39., 40., 41.,
       42., 43., 44., 45., 46., 47., 48., 49., 50., 51., 52., 53., 54.,
       55., 56., 57., 58., 59., 60., 61., 62., 63., 64., 65., 66., 67.,
       68., 69., 70., 71., 72., 73., 74., 75., 76., 77., 78., 79., 80.,
       81., 82., 83., 84., 85., 86., 87., 88., 89., 90., 91., 92., 93.,
       94., 95., 96., 97., 98., 99., nan])

In [141]:
#the simplest operation is to replace the values that are missing with the median of the values: that's because the median is less sensible to the outlies
median_age = customers['age'].median()
median_age

32.0

In [142]:
#in order to prove what I have said before, I calculate also the mean of the age column: we can say that the value is higher
mean_age = customers['age'].mean()
mean_age

36.386964565794

In [143]:
#then I decided to insert the median value of the column where there is the value nan
customers['age'].replace(np.nan, median_age, inplace = True)

In [144]:
#then I decided to check if the substitution went good
#first I decided to print the number of missing values and then the uniques values of the variable
age_values_modified = customers['age'].unique()
age_values_modified.sort()
nan_ages_modified = customers['age'].isna().sum()

print('Unique values of the modified column in the dataset: \n', age_values_modified, '\n')
print('Number of nan values in the modified column in the dataset:', nan_ages_modified)

Unique values of the modified column in the dataset: 
 [16. 17. 18. 19. 20. 21. 22. 23. 24. 25. 26. 27. 28. 29. 30. 31. 32. 33.
 34. 35. 36. 37. 38. 39. 40. 41. 42. 43. 44. 45. 46. 47. 48. 49. 50. 51.
 52. 53. 54. 55. 56. 57. 58. 59. 60. 61. 62. 63. 64. 65. 66. 67. 68. 69.
 70. 71. 72. 73. 74. 75. 76. 77. 78. 79. 80. 81. 82. 83. 84. 85. 86. 87.
 88. 89. 90. 91. 92. 93. 94. 95. 96. 97. 98. 99.] 

Number of nan values in the modified column in the dataset: 0


### FN

In [145]:
#values of the column
values_FN = customers['FN'].unique()
values_FN

array([nan,  1.])

Since the values -1 is not part of the values of the column, I decided to insert the value -1 instead of the value nan

In [146]:
#replacemnt of the nan values
customers['FN'].replace(np.nan, -1, inplace = True)

In [147]:
#verification that everything worked correctly
customers['FN'].unique()

array([-1.,  1.])

### Active

In [148]:
#values of the column
values_active = customers['Active'].unique()
values_active

array([nan,  1.])

In [149]:
#replacemnt of the nan values
customers['Active'].replace(np.nan, -1, inplace = True)

In [150]:
#verification that everything worked correctly
customers['Active'].unique()

array([-1.,  1.])

## Feature enegeneering on a categorical variable: club_member_status

In [151]:
#identification of the values of the variable
values_club_member_status = customers['club_member_status'].unique()
values_club_member_status

array(['ACTIVE', nan, 'PRE-CREATE', 'LEFT CLUB'], dtype=object)

In [152]:
#first: preprocessing on the variable in order to transform the categorical values in ordinal ones
label_encoder = LabelEncoder()
label_encoder.fit(customers['club_member_status'])
results = label_encoder.transform(customers['club_member_status'])
results

array([0, 0, 0, ..., 0, 0, 2])

In [153]:
#creation of a new column with the boolean value True if the customer is active, False in all the other cases
boolean_results = []
for i in range(len(results)):
  if results[i] == 0:
    boolean_results.append(True)
  else:
    boolean_results.append(False)

In [154]:
#control that the boolean column have the same number of active users than the original one
count_active = (customers['club_member_status'] == 'ACTIVE').sum()

count_true = 0
value = True
for i in range(len(boolean_results)):
  if boolean_results[i] == value:
    count_true += 1

print(count_active == count_true)

True


In [155]:
#creation of the new colum in the pandas dataframe with the boolean values
boolean_dataframe = pd.DataFrame(boolean_results)
boolean_dataframe.rename(columns = {0: 'Member status active'}, inplace = True)
boolean_dataframe

Member status active
0                        True
1                        True
2                        True
3                        True
4                        True
...                       ...
1371975                  True
1371976                  True
1371977                  True
1371978                  True
1371979                 False

[1371980 rows x 1 columns]

In [156]:
#merge of the column Customer active with the dataframe customers
customers = pd.concat([customers, boolean_dataframe], axis = 1)
customers

customer_id   FN  Active  \
0        00000dbacae5abe5e23885899a1fa44253a17956c6d1c3... -1.0    -1.0   
1        0000423b00ade91418cceaf3b26c6af3dd342b51fd051e... -1.0    -1.0   
2        000058a12d5b43e67d225668fa1f8d618c13dc232df0ca... -1.0    -1.0   
3        00005ca1c9ed5f5146b52ac8639a40ca9d57aeff4d1bd2... -1.0    -1.0   
4        00006413d8573cd20ed7128e53b7b13819fe5cfc2d801f...  1.0     1.0   
...                                                    ...  ...     ...   
1371975  ffffbbf78b6eaac697a8a5dfbfd2bfa8113ee5b403e474... -1.0    -1.0   
1371976  ffffcd5046a6143d29a04fb8c424ce494a76e5cdf4fab5... -1.0    -1.0   
1371977  ffffcf35913a0bee60e8741cb2b4e78b8a98ee5ff2e6a1...  1.0     1.0   
1371978  ffffd7744cebcf3aca44ae7049d2a94b87074c3d4ffe38...  1.0     1.0   
1371979  ffffd9ac14e89946416d80e791d064701994755c3ab686... -1.0    -1.0   

        club_member_status fashion_news_frequency   age  \
0                   ACTIVE                   NONE  49.0   
1                   ACTIVE                   NONE  25.0   
2                   ACTIVE                   NONE  24.0   
3                   ACTIVE                   NONE  54.0   
4                   ACTIVE              Regularly  52.0   
...                    ...                    ...   ...   
1371975             ACTIVE                   NONE  24.0   
1371976             ACTIVE                   NONE  21.0   
1371977             ACTIVE              Regularly  21.0   
1371978             ACTIVE              Regularly  18.0   
1371979         PRE-CREATE                   NONE  65.0   

                                               postal_code  \
0        52043ee2162cf5aa7ee79974281641c6f11a68d276429a...   
1        2973abc54daa8a5f8ccfe9362140c63247c5eee03f1d93...   
2        64f17e6a330a85798e4998f62d0930d14db8db1c054af6...   
3        5d36574f52495e81f019b680c843c443bd343d5ca5b1c2...   
4        25fa5ddee9aac01b35208d01736e57942317d756b32ddd...   
...                                                    ...   
1371975  7aa399f7e669990daba2d92c577b52237380662f36480b...   
1371976  3f47f1279beb72215f4de557d950e0bfa73789d24acb5e...   
1371977  4563fc79215672cd6a863f2b4bf56b8f898f2d96ed590e...   
1371978  8892c18e9bc3dca6aa4000cb8094fc4b51ee8db2ed14d7...   
1371979  0a1a03306fb2f62164c2a439b38c0caa64b40deaae8687...   

         Member status active  
0                        True  
1                        True  
2                        True  
3                        True  
4                        True  
...                       ...  
1371975                  True  
1371976                  True  
1371977                  True  
1371978                  True  
1371979                 False  

[1371980 rows x 8 columns]

# Export of the new datasets with feature engeneering done in parquet format

In [157]:
#pip install pyarrow

In [158]:
#new_customers.csv file
customers.to_csv('Data/new_customers.csv')